File: 01-intro/03-basic-combinatorial-algorithms/02-variations/only-zero/only-zero.py
class OnlyZero(object):
    elements = []
    elements_count = 0
    base = 2
    ops = [lambda x: x, lambda x: -x]

    def __init__(self, elements: list):
        self.elements = elements
        self.elements_count = len(elements)

    def compute_slot(self, index, slot_index):
        op = (index // (self.base ** slot_index)) % self.base
        return self.ops[op](self.elements[slot_index])

    def format_variation(self, variation):
        output = ""

        for element in variation:
            if element > 0: output += " +{}".format(element)
            else: output += " {}".format(element)

        output += " = 0"

        return output

    def compute(self):
        valid_variations = []
        for index in range(0, self.base ** self.elements_count):
            slot_range = range(0, self.elements_count)
            elements = [self.compute_slot(index, slot_index) for slot_index in slot_range]

            if sum(elements) == 0:
                valid_variations.append(elements)

        return [self.format_variation(variation) for variation in \
                valid_variations]


print(OnlyZero([1, 2, 3, 4, 5, 6, 7, 8]).compute())

File: 01-intro/03-basic-combinatorial-algorithms/02-variations/only-zero/only-zero.php
<?php

throw new Exception("Not working");


class OnlyZero
{
    protected $elements = [];
    protected $elementsCount = 0;
    protected $base = 2;

    public function __construct(array $elements)
    {
        $this->elements = $elements;
        $this->elementsCount = count($elements);
    }

    protected function computeSlot($index, $slotIndex)
    {
        $head = ($this->elementsCount ** $slotIndex);
        $sign = (-1) ** (floor($index / $head) % $this->elementsCount);
        return ($sign * $this->elements[$slotIndex]);
    }

    protected function formatVariation(array $variation)
    {
        $output = "";

        foreach ($variation as $element) {
            if ($element > 0) {
                $output .= " +";
            }
            $output .= $element;
        }

        $output .= " = 0";

        return $output;
    }

    public function compute()
    {
        $validVariations = [];
        foreach (range(0, ($this->base ** $this->elementsCount) - 1) as $index) {
            $slotRange = range(0, $this->elementsCount - 1);
            $elements = [];
            foreach ($slotRange as $slotIndex) {
                $elements[] = $this->computeSlot($index, $slotIndex);
            }

            if (array_sum($elements) == 0) {
                $validVariations[] = $elements;
            }
        }

        return array_map([$this, 'formatVariation'], $validVariations);
    }
}

print_r((new OnlyZero([1, 2, 3, 4, 5, 6, 7, 8]))->compute());